pkg/utils/revision/revision_utils.go

```diff
@@ -0,0 +1,301 @@
+		}
+
+	}
+	return owned, err
+}
+
+func DeepHashObject(hasher hash.Hash, objectToWrite interface{}) {
+	hasher.Reset()
+	printer := spew.ConfigState{
+		Indent:         " ",
+		SortKeys:       true,
+		DisableMethods: true,
+		SpewKeys:       true,
+	}
+	_, err := printer.Fprintf(hasher, "%#v", objectToWrite)
+	if err != nil {
+		return
+	}
+}
+
+func GetRevision(ctx context.Context, k8sClient client.Client, lws *leaderworkerset.LeaderWorkerSet, templateHash string) (*appsv1.ControllerRevision, error) {
+	log := ctrl.LoggerFrom(ctx).WithValues("leaderworkerset", klog.KObj(lws))
+	ctx = ctrl.LoggerInto(ctx, log)
+	if templateHash == "" {
+		return nil, nil
+	}
+	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{MatchLabels: map[string]string{
+		leaderworkerset.TemplateRevisionHashKey: templateHash,
+	}})
+	if err != nil {
+		return nil, err
+	}
+	revisions, err := ListRevisions(ctx, k8sClient, lws, selector)
+	if err != nil {
+		log.Error(err, "Listing all controller revisions")
+		return nil, err
+	}
+
+	if len(revisions) == 0 {
+		return nil, nil
+	}
+
+	if len(revisions) > 1 {
+		// Since we only create a controllerRevision when the template hash changes, only one should match
+		log.Error(err, "More than one revision exists for the given template hash; returning the latest revision")
+		return revisions[len(revisions)-1], nil
+	}
+
+	return revisions[0], nil
+}
+
+// GetPatch returns a strategic merge patch that can be applied to restore a LeaderWorkerSet to a
+// previous version. If the returned error is nil the patch is valid. The current state that we save is the
+// leaderWorkerTemplate and NetworkConfig. We can modify this later to encompass more state (or less) and
+// remain compatible with previously recorded patches.
+func GetPatch(lws *leaderworkerset.LeaderWorkerSet) ([]byte, error) {
+	str := &bytes.Buffer{}
+	clone := lws.DeepCopy()
+	// When upgrading from an LWS version that doesn't contain NetworkConfig, NetworkConfig will be nil
+	// until another field in the LWS object is changed triggering the LWS webhook. This allows the revision
+	// to be the same before and after the LWS webhook actually defaults the value.
+	if clone.Spec.NetworkConfig == nil {
+		clone.Spec.NetworkConfig = &leaderworkerset.NetworkConfig{}
+		subdomainPolicy := leaderworkerset.SubdomainShared
+		clone.Spec.NetworkConfig = &leaderworkerset.NetworkConfig{
+			SubdomainPolicy: &subdomainPolicy,
+		}
+	}
+
+	if err := unstructured.UnstructuredJSONScheme.Encode(clone, str); err != nil {
+		return nil, err
+	}
+	var raw map[string]interface{}
+	if err := json.Unmarshal(str.Bytes(), &raw); err != nil {
+		return nil, err
+	}
+	objCopy := make(map[string]interface{})
+	specCopy := make(map[string]interface{})
+	spec := raw["spec"].(map[string]interface{})
+	specCopy["networkConfig"] = spec["networkConfig"]
+	specCopy["leaderWorkerTemplate"] = spec["leaderWorkerTemplate"].(map[string]interface{})
+	specCopy["$patch"] = "replace"
```

Comment:
I think this will cause the whole spec to be replaced, what we want is to replace the `leaderWorkerTemplate` and `networkConfig` fields only, the rest should stay the same.

The unit test can catch this: try to set a field in the incoming LWS spec other than networkConfig and leaderWorkerTemplate, the restored lws should have the field, I think now it would not.

Reply:
Yes, it does replace the whole spec, sets the other spec fields to nil. Fixed and changed unit test to account for it.